### PR TITLE
Revision deployment approval workflow

### PR DESCRIPTION
**Pull request by GihanAyesh**
Front end code for revision deployment workflow

Git issue: https://github.com/wso2/api-manager/issues/2137

### CODE REVIEW

File: portals/admin/src/main/webapp/site/public/locales/en.json [hunk 1]
```diff
@@ -614,6 +600,25 @@
   "Workflow.APIProductStateChange.List.empty.title.apistatechange": "API Product State Change",
   "Workflow.APIProductStateChange.apicall.has.errors": "Unable to get workflow pending requests for API Product State Change",
   "Workflow.APIProductStateChange.updateStatus.has.errors": "Unable to complete API Product state change approve/reject process.",
+  "Workflow.APIRevisionDeployment.List.empty.content.revisiondeployments": "There are no pending workflow requests for revision deployment.",
+  "Workflow.APIRevisionDeployment.List.empty.title.revisiondeployments": "Revision Deployment",
+  "Workflow.APIRevisionDeployment.apiCall.has.errors": "Unable to get workflow pending requests for Revision Deployment",
+  "Workflow.APIRevisionDeployment.help.link.one": "Create a Revision Deployment approval workflow Request",
+  "Workflow.APIRevisionDeployment.permission.denied.content": "You dont have enough permission to view Revision Deployment - Approval Tasks. Please contact the site administrator.",
```
Comment (tharikaGitHub) [inline portals/admin/src/main/webapp/site/public/locales/en.json:607]:
```suggestion
  "Workflow.APIRevisionDeployment.permission.denied.content": "You don't have enough permission to view Revision Deployment - Approval Tasks. Please contact the site administrator.",
```

File: portals/admin/src/main/webapp/source/src/app/components/Workflow/APIRevisionDeployment.jsx [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+
+    /**
+     * API call to get Detected Data
+     * @returns {Promise}.
+     */
+    function apiCall() {
+        return restApi
+            .workflowsGet('AM_REVISION_DEPLOYMENT')
+            .then((result) => {
+                const workflowlist = result.body.list.map((obj) => {
```
Comment (tharikaGitHub) [inline portals/admin/src/main/webapp/source/src/app/components/Workflow/APIRevisionDeployment.jsx:103]: Can obj and obj.properties be null?

File: portals/admin/src/main/webapp/source/src/app/components/Workflow/APIRevisionDeployment.jsx [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+                <List component='nav' aria-label='main mailbox folders'>
+                    <ListItem button>
+                        <ListItemIcon>
+                            <DescriptionIcon />
+                        </ListItemIcon>
+                        <Link
+                            target='_blank'
+                            href={Configurations.app.docUrl
+                        + 'learn/consume-api/manage-application/advanced-topics/'
+                        + 'adding-an-application-creation-workflow/#adding-an-application-creation-workflow'}
```
Comment (tharikaGitHub) [inline portals/admin/src/main/webapp/source/src/app/components/Workflow/APIRevisionDeployment.jsx:201]: Let's update this link when the doc PR is merged.

File: portals/admin/src/main/webapp/source/src/app/components/Workflow/APIRevisionDeployment.jsx [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+                title={(
+                    <FormattedMessage
+                        id='Workflow.APIRevisionDeployment.permission.denied.title'
+                        defaultMessage='Permission Denied'
+                    />
+                )}
+                content={(
+                    <FormattedMessage
+                        id='Workflow.APIRevisionDeployment.permission.denied.content'
+                        defaultMessage={'You dont have enough permission to view Revision Deployment - '
```
Comment (tharikaGitHub) [inline portals/admin/src/main/webapp/source/src/app/components/Workflow/APIRevisionDeployment.jsx:438]:
```suggestion
                        defaultMessage={'You don't have enough permission to view Revision Deployment - '
```

File: portals/publisher/src/main/webapp/source/src/app/components/Apis/Details/Environments/Environments.jsx [hunk 5]
```diff
@@ -764,31 +764,59 @@ export default function Environments() {
...
+                        Alert.info('Deploy revision successfully');
+                    } else if ( response.obj[0]?.status === 'CREATED'){
+                        Alert.info('Deploy revision request sent successfully');
+                    }
+                }
+            }).catch((error) => {
+                if (error.response) {
+                    Alert.error(error.response.body.description);
+                } else {
+                    Alert.error('Something went wrong while deploy the revision');
```
Comment (tharikaGitHub) [inline portals/publisher/src/main/webapp/source/src/app/components/Apis/Details/Environments/Environments.jsx:813]:
```suggestion
                    Alert.error('Something went wrong while deploying the revision');
```

File: portals/publisher/src/main/webapp/source/src/app/components/Apis/Details/Environments/Environments.jsx [hunk 6]
```diff
@@ -1483,6 +1511,246 @@ export default function Environments() {
...
+                    id='Apis.Details.Environments.Environments.status.not.deployed'
+                    defaultMessage='No Revision Deployed'
+                />
+            );
+            
+        }
+
+        const approvedDeployment =gatewayRevisions.find(revision => {
+            return revision.deploymentInfo.some(info => (info.status === null || info.status === "APPROVED") && 
+            info.name === row.name);});
```
Comment (tharikaGitHub) [inline portals/publisher/src/main/webapp/source/src/app/components/Apis/Details/Environments/Environments.jsx:1719]: Let's fix the formatting here for readability.

File: portals/publisher/src/main/webapp/source/src/app/components/Apis/Details/Environments/Environments.jsx [hunk 7]
```diff
@@ -1730,6 +1999,10 @@ export default function Environments() {
                                     {internalGateways && internalGateways.map((row) => (
                                         <Grid item xs={4}>
                                             <Card
+                                                style={{ opacity: allEnvRevision.some(o1 => {
+                                                    return o1.deploymentInfo.some(o2 => 
+                                                        o2.name === row.name && 
+                                                        o2.status === 'CREATED');})? '0.6' : '1' }}
```
Comment (tharikaGitHub) [inline portals/publisher/src/main/webapp/source/src/app/components/Apis/Details/Environments/Environments.jsx:2005]: Let's fix the formatting here too for readability.

File: portals/publisher/src/main/webapp/source/src/app/components/Apis/Details/Environments/Environments.jsx [hunk 8]
```diff
@@ -1738,19 +2011,40 @@ export default function Environments() {
...
-                                                                inputProps={{ 'aria-label': 'secondary checkbox' }}
-                                                                data-testid={row.displayName + 'gateway-select-btn'}
-                                                            />
-                                                        )}
+                                                        action={
+                                                            <>
+                                                                {allEnvRevision.some(o1 => {
+                                                                    return o1.deploymentInfo.some(o2 => 
+                                                                        o2.name === row.name && 
+                                                                        o2.status === 'CREATED');})?
```
Comment (tharikaGitHub) [inline portals/publisher/src/main/webapp/source/src/app/components/Apis/Details/Environments/Environments.jsx:2019]: Formatting

File: portals/publisher/src/main/webapp/source/src/app/components/Apis/Details/Environments/Environments.jsx (hunk 7 above)
Comment (GihanAyesh) [reply portals/publisher/src/main/webapp/source/src/app/components/Apis/Details/Environments/Environments.jsx]: fixed with https://github.com/wso2/apim-apps/pull/509/commits/feb7a470cc8167eb79976df580ae001188e69a10

File: portals/publisher/src/main/webapp/source/src/app/components/Apis/Details/Environments/Environments.jsx (hunk 6 above)
Comment (GihanAyesh) [reply portals/publisher/src/main/webapp/source/src/app/components/Apis/Details/Environments/Environments.jsx]: fixed with https://github.com/wso2/apim-apps/commit/feb7a470cc8167eb79976df580ae001188e69a10

File: portals/admin/src/main/webapp/source/src/app/components/Workflow/APIRevisionDeployment.jsx (hunk 2 above)
Comment (GihanAyesh) [reply portals/admin/src/main/webapp/source/src/app/components/Workflow/APIRevisionDeployment.jsx:103]: This obj would not be empty as if there are no revisions the results body is empty and if obj are available they always have properties.

**Review by tharikaGitHub — APPROVED** [review]

**Pull request merged**